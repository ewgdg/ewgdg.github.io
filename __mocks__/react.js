const React = jest.requireActual("react")
module.exports = {
  ...React,
  useContext: jest.fn().mockImplementation(React.useContext),
}
