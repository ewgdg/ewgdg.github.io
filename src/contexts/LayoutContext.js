import React from "react"

const contextValueRef = { current: { scrollLayer: null, historyState: {} } }
const context = React.createContext(contextValueRef.current)
export default context
export { contextValueRef }
