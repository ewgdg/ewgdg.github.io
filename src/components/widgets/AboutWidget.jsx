/* eslint-disable react/destructuring-assignment */
/* eslint-disable react/jsx-one-expression-per-line */
/* eslint-disable import/no-unresolved */
/* eslint-disable new-cap */
import React, { useRef, useEffect, useContext } from "react"
import { TimelineLite, Power3, Power1 } from "gsap"

import { makeStyles } from "@material-ui/styles"
import { Link } from "gatsby"
import LayoutContext from "../../contexts/LayoutContext"

import CharSequence from "../sections/CharSequence"
import { ScrollDetector } from "../../utils/scroll"
import FlexContainer from "../sections/FlexContainer"

const useStyles = makeStyles({
  charSequenceContainer: {
    height: "4rem",
    width: "100%",
    fontSize: "3.5rem",
    position: "relative",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    paddingLeft: "20%",
  },
  charSequenceBackground: {
    backgroundImage:
      "linear-gradient(130deg, #a7fcf5 15%, #1fc8cb 50%, #2cb5d8 75%)",
    position: "absolute",
    top: 0,
    left: 0,
    width: "120%",
    height: "100%",
    zIndex: -1,
    opacity: 0.6,
  },
  ribbon: {
    width: "100%",
    height: "350px",
    backgroundColor: "rgb(255, 32, 126)",
    opacity: 0.6,
  },
  skewed: {
    transform: "skewY(-15deg)",
  },
  textContainerInsideRibbon: {
    position: "absolute",
    top: 0,
    left: 0,
    width: "100%",
    height: "100%",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    flexDirection: "column",
    padding: "0 10% 0 10%",
  },
})

function About() {
  const ribbonContainerRef = useRef(null)
  const ribbonRef = useRef(null)
  const charSequenceBackgroundRef = useRef(null)
  const charRefs = useRef([])
  const context = useContext(LayoutContext)
  if (!context.scrollLayer) return null
  const classes = useStyles()
  // add scroll animation controll
  useEffect(() => {
    const animation = new TimelineLite()
    // animate ribbon
    animation
      .add("startAnimation")
      .from(ribbonRef.current, 0.7, {
        scaleX: 0,
        transformOrigin: "right",
        ease: Power3.easeInOut,
      })
      .add("ribbonFinished")
      .from(charSequenceBackgroundRef.current, 0.7, {
        scaleX: 0,
        transformOrigin: "left",
        ease: Power3.easeInOut,
      })

    animation
      .staggerTo(
        charRefs.current,
        0.15,
        {
          fontSize: "2em",
          stagger: 0.1,
          ease: Power1.easeInOut,
        },
        undefined,
        "startAnimation+=0.2"
      )
      .staggerTo(
        charRefs.current,
        0.15,
        {
          fontSize: "1em",
          stagger: 0.1,
          ease: Power1.easeInOut,
        },
        undefined,
        "startAnimation+=0.35"
      )
      .staggerFrom(
        ".animatedline",
        0.1,
        {
          opacity: 0,
          stagger: 0.05,
        },
        undefined,
        "ribbonFinished"
      )
      .add("endAnimation")

    animation.pause()

    const scrollDetector = new ScrollDetector({
      scrollLayer: context.scrollLayer,
      triggerElement: ribbonContainerRef.current,
      triggerHook: 0.5,
      duration: 0,
      throttleLimit: 0,
    })
    scrollDetector.setEventListener(progress => {
      if (progress > 0) {
        animation.play()
      } else {
        animation.reverse()
      }
    })
    scrollDetector.update()

    return () => {
      animation.progress(1)
      animation.kill()

      scrollDetector.destroy()
    }
  }, [context])

  return (
    <FlexContainer>
      <div style={{ width: "100%", height: "auto" }}>
        <div style={{ display: "inline-block" }}>
          <div className={classes.charSequenceContainer}>
            <CharSequence string="About" charRefs={charRefs} />
            <div
              ref={charSequenceBackgroundRef}
              className={classes.charSequenceBackground}
            />
          </div>
        </div>
        <div style={{ position: "relative" }}>
          <div
            id="ribbonContainer"
            ref={ribbonContainerRef}
            className={classes.skewed}
          >
            <div id="ribbon" ref={ribbonRef} className={classes.ribbon} />
          </div>
          <div className={classes.textContainerInsideRibbon}>
            <div id="animatedLineGroup">
              <h2 className="animatedline">
                My passion is about solving challenging problem.
              </h2>
              <div>
                <div className="animatedline">
                  And that is{" "}
                  <Link to="/blog/2019-10-30-why-i-created-this-page">why</Link>{" "}
                  I made this page.
                </div>
                <div className="animatedline">
                  I like technology and computer programming and many more.
                </div>
                <div className="animatedline" style={{ opacity: 0.8 }}>
                  If you want to learn more about me, keep{" "}
                  <Link to="/about">reading</Link>.
                </div>
                <div className="animatedline" style={{ opacity: 0.6 }}>
                  Or send me an{" "}
                  <button
                    style={{
                      background: "none",
                      border: "none",
                      padding: "0",
                      color: "#069",
                      textDecoration: "underline",
                      cursor: "pointer",
                    }}
                    type="button"
                    onClick={() => {
                      window.location.href = "m&a&i#l#&t#o:$x#&i$a$n&#$.$z$#5$1#&$2$#^$#g$m#a$i&$l#$.$c#o&#m$"
                        .replace(/\^/g, "@")
                        .replace(/[#&$]/g, "")
                    }}
                  >
                    email.
                  </button>
                </div>
                <div className="animatedline" style={{ opacity: 0.4 }}>
                  ... ...
                </div>
                <div className="animatedline" style={{ opacity: 0.2 }}>
                  ...
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </FlexContainer>
  )
}

export default About
