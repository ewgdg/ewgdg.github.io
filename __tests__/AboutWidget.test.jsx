/* eslint-disable react/jsx-filename-extension */
import { render } from "@testing-library/react"
import React from "react"
import AboutWidget from "../src/components/widgets/AboutWidget"
import LayoutContext from "../src/contexts/LayoutContext"

const { useContext } = jest.requireActual("react")

describe("About Widget", () => {
  beforeAll(() => {
    // set context
    const { getByTestId } = render(<div data-testid="scrollLayer" />)
    React.useContext.mockImplementation(context => {
      if (Object.is(LayoutContext, context)) {
        return {
          scrollLayer: getByTestId("scrollLayer"),
        }
      }
      return useContext(context)
    })
  })

  it("render properly", () => {
    const { container } = render(<AboutWidget />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
