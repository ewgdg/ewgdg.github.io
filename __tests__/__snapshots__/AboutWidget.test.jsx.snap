// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`About Widget render properly 1`] = `
<div
  class="makeStyles-flexContainer-6 undefined"
>
  <div
    style="width: 100%; height: auto;"
  >
    <div
      style="display: inline-block;"
    >
      <div
        class="makeStyles-charSequenceContainer-1"
      >
        <span
          style="font-size: 1em;"
        >
          A
        </span>
        <span
          style="font-size: 1em;"
        >
          b
        </span>
        <span
          style="font-size: 1em;"
        >
          o
        </span>
        <span
          style="font-size: 1em;"
        >
          u
        </span>
        <span
          style="font-size: 1em;"
        >
          t
        </span>
        <div
          class="makeStyles-charSequenceBackground-2"
          style="transform-origin: left; transform: matrix(0,0,0,1,0,0);"
        />
      </div>
    </div>
    <div
      style="position: relative;"
    >
      <div
        class="makeStyles-skewed-4"
        id="ribbonContainer"
      >
        <div
          class="makeStyles-ribbon-3"
          id="ribbon"
          style="transform-origin: right; transform: matrix(0,0,0,1,0,0);"
        />
      </div>
      <div
        class="makeStyles-textContainerInsideRibbon-5"
      >
        <div
          id="animatedLineGroup"
        >
          <h2
            class="animatedline"
            style="opacity: 0;"
          >
            My passion is about solving challenging problem.
          </h2>
          <div>
            <div
              class="animatedline"
              style="opacity: 0;"
            >
              And that is
               
              <a
                href="/blog/2019-10-30-why-i-created-this-page"
              >
                why
              </a>
               
              I made this page.
            </div>
            <div
              class="animatedline"
              style="opacity: 0;"
            >
              I like technology and computer programming and many more.
            </div>
            <div
              class="animatedline"
              style="opacity: 0;"
            >
              If you want to learn more about me, keep
               
              <a
                href="/about"
              >
                reading
              </a>
              .
            </div>
            <div
              class="animatedline"
              style="opacity: 0;"
            >
              Or send me an
               
              <button
                style="background: none; padding: 0px; color: rgb(0, 102, 153); text-decoration: underline; cursor: pointer;"
                type="button"
              >
                email.
              </button>
            </div>
            <div
              class="animatedline"
              style="opacity: 0;"
            >
              ... ...
            </div>
            <div
              class="animatedline"
              style="opacity: 0;"
            >
              ...
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
